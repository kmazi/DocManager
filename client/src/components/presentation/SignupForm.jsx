import React from 'react';
import PropTypes from 'prop-types';
import Alert from 'sweetalert2';
import $ from 'jquery';

/**
 * Fires the function that authenticates a user
 * @param {object} event - object containing data about the control that
 * triggered an event
 * @param {func} signUserUp - function that executes when signing users up
 * @param {object} history - the browser and app state history
 * @param {oobject} roleType - The role a user belongs to
 * @param {func} allDocuments - Function that executes when
   * fetching documents accessible to a particular user
 * @return {null} returns void
 */
const signUp = (event, signUserUp, history, roleType, allDocuments) => {
  event.preventDefault();
  const userName = $('#signupform input[name=username]').val();
  const email = $('#signupform input[type=email]').val();
  const password = $('#signupform input[name=password]').val();
  const comfirmPassword = $('#signupform input[name=comfirmpassword]').val();
  const roleValue = $('#signupform input[name=group1]:checked')
    .attr('data-value');
  const roleId = $('#signupform input[name=group1]:checked').val();
  const isactive = true;
  const formData =
    { userName, email, password, comfirmPassword, roleId, roleValue, isactive };
  if (password === comfirmPassword) {
    signUserUp(formData)
      .then((res) => {
        if (res.status === 'successful') {
          allDocuments(roleType);
          history.push('/user/documents');
        } else {
          Alert({
            title: 'Error Signing up',
            text: res.message,
            type: 'error',
            confirmButtonText: 'ok'
          });
        }
      });
  } else {
    Alert({
      title: 'Password mispatch',
      text: 'Ensure your password is spelt correctly',
      type: 'error',
      confirmButtonText: 'ok'
    });
  }
};
/**
 * Renders form for signin up new user
 * @return {object} Returns the signup form to render
 */
const SignupForm = ({ signUserUp, history,
  submitButton, roleType, allDocuments }) => (
    <div id="signupform" style={{ display: 'none' }}>
      <input
        className="forminput"
        name="username"
        type="text"
        minLength="3"
        placeholder="Username"
        required
      />
      <input
        className="forminput"
        name="email"
        type="email"
        minLength="10"
        placeholder="Email"
        required
      />
      <input
        className="forminput"
        name="password"
        type="password"
        minLength="6"
        placeholder="Password"
        required
      />
      <input
        className="forminput"
        name="comfirmpassword"
        type="password"
        placeholder="Comfirm password"
        required
      />
      <div id="roles">
        Role:&nbsp;&nbsp;
        <input
          className="with-gap"
          name="group1"
          type="radio"
          data-value="Learning"
          id="Learning"
          value="4"
        />
        <label htmlFor="Learning">Learning</label>
        <input
          className="with-gap"
          name="group1"
          type="radio"
          data-value="Fellow"
          id="Fellow"
          value="3"
        />
        <label id="Fellowlabel" htmlFor="Fellow">Fellow</label>
        <input
          className="with-gap"
          name="group1"
          type="radio"
          data-value="DevOps"
          value="5"
          id="DevOps"
        />
        <label htmlFor="DevOps">DevOps</label><br />
      </div>
      <button
        id="signup"
        className="center-align waves-effect waves-light btn"
        onClick={
          event => signUp(event, signUserUp, history, roleType, allDocuments)}
      >{submitButton}</button>
    </div>
  );

SignupForm.propTypes = {
  signUserUp: PropTypes.func.isRequired,
  roleType: PropTypes.string.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
  submitButton: PropTypes.string.isRequired,
  allDocuments: PropTypes.func.isRequired,
};
export default SignupForm;
