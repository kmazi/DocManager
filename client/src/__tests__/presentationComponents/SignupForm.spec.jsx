import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import SignupForm from '../../components/presentation/SignupForm';

describe('The signupform component:', () => {
  const history = { push: url => url };
  const response = { status: 'successful' };
  const props = {
    signUserUp: jest.fn(() => Promise.resolve(response)),
    allDocuments: jest.fn(() => Promise.resolve()),
    submitButton: 'submit',
    roleType: 'Admin',
    history,
  };
  test('should render the signup form correctly', () => {
    const component = renderer.create(
      <SignupForm {...props} />
  );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test(`should submit the form to signup a user when signupbtn is clicked
  and the form is valid`, () => {
    const event = {
      preventDefault: () => 'hello',
    };
    const enzymeWrapper = mount(<SignupForm {...props} />);
    enzymeWrapper.find('button#signup').props().onClick(event);
    expect(props.signUserUp.mock.calls.length).toBe(1);
  });

  test('should alert an error message when signup fails', () => {
    response.status = 'unsuccessful';
    const event = {
      preventDefault: () => 'hello',
    };
    const enzymeWrapper = mount(<SignupForm {...props} />);
    enzymeWrapper.find('button#signup').props().onClick(event);
    const component = enzymeWrapper;
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });
});
