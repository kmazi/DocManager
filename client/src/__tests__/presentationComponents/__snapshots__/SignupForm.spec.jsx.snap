// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The signupform component: should alert an error message when signup fails 1`] = `
<SignupForm
  allDocuments={[Function]}
  history={
    Object {
      "push": [Function],
    }
  }
  roleType="Admin"
  signUserUp={[Function]}
  submitButton="submit"
>
  <div
    id="signupform"
    style={
      Object {
        "display": "none",
      }
    }
  >
    <input
      className="forminput"
      minLength="3"
      name="username"
      placeholder="Username"
      required={true}
      type="text"
    />
    <input
      className="forminput"
      minLength="10"
      name="email"
      placeholder="Email"
      required={true}
      type="email"
    />
    <input
      className="forminput"
      minLength="6"
      name="password"
      placeholder="Password"
      required={true}
      type="password"
    />
    <input
      className="forminput"
      name="comfirmpassword"
      placeholder="Comfirm password"
      required={true}
      type="password"
    />
    <div
      id="roles"
    >
      Role:  
      <input
        className="with-gap"
        data-value="Learning"
        id="Learning"
        name="group1"
        type="radio"
        value="4"
      />
      <label
        htmlFor="Learning"
      >
        Learning
      </label>
      <input
        className="with-gap"
        data-value="Fellow"
        id="Fellow"
        name="group1"
        type="radio"
        value="3"
      />
      <label
        htmlFor="Fellow"
        id="Fellowlabel"
      >
        Fellow
      </label>
      <input
        className="with-gap"
        data-value="DevOps"
        id="DevOps"
        name="group1"
        type="radio"
        value="5"
      />
      <label
        htmlFor="DevOps"
      >
        DevOps
      </label>
      <br />
    </div>
    <button
      className="center-align waves-effect waves-light btn"
      id="signup"
      onClick={[Function]}
    >
      submit
    </button>
  </div>
</SignupForm>
`;

exports[`The signupform component: should render the signup form correctly 1`] = `
<div
  id="signupform"
  style={
    Object {
      "display": "none",
    }
  }
>
  <input
    className="forminput"
    minLength="3"
    name="username"
    placeholder="Username"
    required={true}
    type="text"
  />
  <input
    className="forminput"
    minLength="10"
    name="email"
    placeholder="Email"
    required={true}
    type="email"
  />
  <input
    className="forminput"
    minLength="6"
    name="password"
    placeholder="Password"
    required={true}
    type="password"
  />
  <input
    className="forminput"
    name="comfirmpassword"
    placeholder="Comfirm password"
    required={true}
    type="password"
  />
  <div
    id="roles"
  >
    Role:  
    <input
      className="with-gap"
      data-value="Learning"
      id="Learning"
      name="group1"
      type="radio"
      value="4"
    />
    <label
      htmlFor="Learning"
    >
      Learning
    </label>
    <input
      className="with-gap"
      data-value="Fellow"
      id="Fellow"
      name="group1"
      type="radio"
      value="3"
    />
    <label
      htmlFor="Fellow"
      id="Fellowlabel"
    >
      Fellow
    </label>
    <input
      className="with-gap"
      data-value="DevOps"
      id="DevOps"
      name="group1"
      type="radio"
      value="5"
    />
    <label
      htmlFor="DevOps"
    >
      DevOps
    </label>
    <br />
  </div>
  <button
    className="center-align waves-effect waves-light btn"
    id="signup"
    onClick={[Function]}
  >
    submit
  </button>
</div>
`;
